File: Binary Tree/106M. Construct Binary Tree from Inorder and Postorder Traversal.cpp
/*
106M. Construct Binary Tree from Inorder and Postorder Traversal
Tree, Array, Depth-first Search

Desciption:
Given two integer arrays inorder and postorder where inorder is the inorder traversal of a binary tree and postorder is the postorder traversal of the same tree, construct and return the binary tree.

Constraints:
1. inorder and postorder consist of unique values.
2. Each value of postorder also appears in inorder.
3. inorder is guaranteed to be the inorder traversal of the tree.
4. postorder is guaranteed to be the postorder traversal of the tree.

Example 1:
Input: inorder = [9,3,15,20,7], postorder = [9,15,7,20,3]
Output: [3,9,20,null,null,15,7]
          3
      9       20
           15    7

Example 2:
Input: inorder = [-1], postorder = [-1]
Output: [-1]

Similar Quetion:
Construct Binary Tree from Preorder and Inorder Traversal - Medium
*/

/*
Method: The idea is similar to 105M.
        The structure of inorder is L - C - R, the structure of postorder is L - R - C.
        We can use postorder array to determine the root value, and use inorder array to determine the subtrees.
*/
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {
        return build(inorder, 0, inorder.size()-1, postorder, 0, postorder.size()-1);
    }
    
    TreeNode* build(vector<int>& inorder, int inStart, int inEnd, vector<int>& postorder, int postStart, int postEnd){
        if (inStart > inEnd){
            return NULL;
        }
        int rootVal = postorder[postEnd], index;
        for(int i=inStart;i<=inEnd;i++){
            if(inorder[i]==rootVal){
                index = i;
                break;
            }
        }
        TreeNode* root = new TreeNode(rootVal);
        int length_left = index - inStart;
        root->left = build(inorder, inStart, index-1, postorder, postStart, postStart+length_left-1);
        root->right = build(inorder, index+1, inEnd, postorder, postStart+length_left, postEnd-1);
        return root;
    }
};

File: Binary Tree/106M. Construct Binary Tree from Inorder and Postorder Traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def buildTree(self, inorder, postorder):
        """
        :type inorder: List[int]
        :type postorder: List[int]
        :rtype: TreeNode
        """
        return self.build(inorder, 0, len(inorder)-1, postorder, 0, len(postorder)-1)
        
    def build(self, inorder, inStart, inEnd, postorder, postStart, postEnd):
        if inStart > inEnd:
            return None
        rootVal = postorder[postEnd]
        for i in range(inStart, inEnd+1):
            if inorder[i]==rootVal:
                index = i
                break
        root = TreeNode(rootVal)
        length_left = index - inStart
        root.left = self.build(inorder, inStart, index-1, postorder, postStart, postStart+length_left-1)
        root.right = self.build(inorder, index+1, inEnd, postorder, postStart+length_left, postEnd-1)
        return root
        
        
